dvc/remote/base.py

```diff
@@ -731,22 +740,32 @@ def all(self, jobs=None, name=None):
             remote_size, remote_checksums, jobs, name
         )
 
-    def gc(self, named_cache, jobs=None):
-        logger.debug("named_cache: {} jobs: {}".format(named_cache, jobs))
-        used = self.extract_used_local_checksums(named_cache)
+    def gc(self, named_caches, jobs=None):
+        used = self.extract_used_local_checksums(named_caches)
 
         if self.scheme != "":
-            used.update(named_cache[self.scheme])
+            for dir_cache, file_cache in named_caches:
+                if dir_cache:
+                    used.update(dir_cache[self.scheme])
+                used.update(file_cache[self.scheme])
 
         removed = False
-        for checksum in self.all(jobs, str(self.path_info)):
+        # checksums must be sorted to ensure we always remove .dir files first
+        for checksum in sorted(
+            self.all(jobs, str(self.path_info)),
+            key=self.is_dir_checksum,
+            reverse=True,
+        ):
```

Comment:
This is another place we would benefit from having separated .dir and file checksums in our cache structure